libbpf-tools/vfsstat.bpf.c

```diff
@@ -15,32 +15,32 @@ static __always_inline int inc_stats(int key)
 	return 0;
 }
 
-SEC("kprobe/vfs_read")
-int BPF_KPROBE(vfs_read)
+SEC("fentry/vfs_read")
```

Comment:
using fentry/fexit bumps the required minimal Linux version, so it's a regression in the behavior in that sense. Best way to get better performance by using fentry/fexit entries would be to have both kprobe/kretprobe and fentry/fexit and then detect fentry/fexit support at runtime and load (or rather disable auto-load of) only appropriate set of programs

Reply:
Although the minimum required kernel version was not explicitly stated in README,  I thought we could use BPF trampoline  without considering older kernels.
Actually, fentry/fexit is used in tcpconnlat without considering older kernels:
https://github.com/iovisor/bcc/blob/v0.18.0/libbpf-tools/tcpconnlat.bpf.c#L49

What is the minimum kernel version required to use libbpf-tools?
If the minimum kernel version is 5.5 or later, I think that we do not need to support kprobe/kretprobe for vfsstat.

Reply:
There is no single minimal kernel version for all libbpf-tools. vfsstat in particular is extremely simple and doesn't use CO-RE, so it can run on very old kernels. It's a good property, so we should try to preserve it. This tool could be a good example on how to opportunistically use fentry/fexit (for better performance), but fallback to kprobe/kretprobe (for wider kernel support).

Reply:
Thank you for your review.
In addition to that, I remembered that BPF trampoline was only supported on x86_64 currently.
So the fallback to kprobe/kretprobe is needed for not only older kernels support but also multi-architecture support.
I'll try on it.

Reply:
@anakryiko 
I'm sorry for the late reply.

I tried to implement fallback to kprobe/kretprobe but it was needed to prepare for two skel files which are for kprobes and fentry.
In addition, vfsstat.c needed to implement the abstraction layer to use different tracing mechanism depending on the kernel version and architecture.
I think that it is redundant for all tools to implement same facilities.
So libbpf may provide new APIs to realize fallback. Or, Could you give me your idea if you have?
Anyway, I'll continue this activity when I have some time but chnage this PR to draft.

Reply:
You shouldn't need two skeletons. Please see recent ext4dist PR on how it is dealing with having both kprobe and fentry functions (https://github.com/iovisor/bcc/pull/3309).

I think before we add some extra functionality to libbpf, let's just first build up libbpf-tools infra to be used by libbpf-tools for starters (and to understand what's essential and generalizable vs what's just specific to these tools). How about we implement feature-probing helpers for common scenarios, just like BPF.support_kfunc in Python side of BCC:

```
 908     @staticmethod¬
 909     def support_kfunc():¬
 910         # there's no trampoline support for other than x86_64 arch¬
 911         if platform.machine() != 'x86_64':¬
 912             return False;¬
 913         if not lib.bpf_has_kernel_btf():¬
 914             return False;¬
 915         # kernel symbol "bpf_trampoline_link_prog" indicates kfunc support¬
 916         if BPF.ksymname("bpf_trampoline_link_prog") != -1:¬
 917             return True¬
 918         return False¬
```

Checking kernel BTF is simple (`access("/sys/kernel/btf/vmlinux", F_OK);`). We probably need a fast implementation for checking the existence of a single (or maybe small set of) kernel symbols. See trace_helpers.c, it has efficient ksyms cache implementation, but that's overkill if you only need to check existence of a symbol. platform.machine() is probably implementable with utsname() call? WDYT?

Reply:
Thank you for your valuable information.

>You shouldn't need two skeletons. Please see recent ext4dist PR on how it is dealing with having both kprobe and fentry functions (#3309).

I could not execute `ext4dist` in my environment without BTF.
I think that libbpf checks the existence of BTF when it actually loads the skel object using bpf_object__load_skeleton().

I just glanced at https://github.com/libbpf/libbpf/blob/99bc176337071326d7d6b9552142eb6a0450fc88/src/libbpf.c and I found that libbpf checks the existence of BTF at startup if the bpf object needs BTF.
Below code checks whether the bpf object needs BTF	
https://github.com/libbpf/libbpf/blob/99bc176337071326d7d6b9552142eb6a0450fc88/src/libbpf.c#L2588-L2589
For example, above code will not return 0 and execute libbpf_find_kernel_btf() and fail to find BTF if I change `vfsstat.bpf.c` to use fentry. 
So I could not include both SEC() of fentry and kprobes in `vfsstat.bpf.c`.

I apologize if there are any mistakes in my understanding.
I'll investigate it more.

>I think before we add some extra functionality to libbpf, let's just first build up libbpf-tools infra to be used by libbpf-tools for starters (and to understand what's essential and generalizable vs what's just specific to these tools). How about we implement feature-probing helpers for common scenarios, just like BPF.support_kfunc in Python side of BCC:

I agree with your idea.

>Checking kernel BTF is simple (access("/sys/kernel/btf/vmlinux", F_OK);). We probably need a fast implementation for checking the existence of a single (or maybe small set of) kernel symbols. See trace_helpers.c, it has efficient ksyms cache implementation, but that's overkill if you only need to check existence of a symbol. platform.machine() is probably implementable with utsname() call? WDYT?

I'll read trace_helpers.c at first.
Thanks!

Reply:
>For example, above code will not return 0 and execute libbpf_find_kernel_btf() and fail to find BTF if I change vfsstat.bpf.c to use fentry.

I understand the mechanism of bpf_program__set_autoload() and it bypasses the check at runtime when false.
But below code may not be correct. I think that `if (!should_fallback()) {` is correct instead of `if (should_fallback()) {` 
https://github.com/iovisor/bcc/blob/7fd8cad30f910af0470e00130817442091376ab3/libbpf-tools/ext4dist.c#L195

Reply:
> For example, above code will not return 0 and execute libbpf_find_kernel_btf() and fail to find BTF if I change `vfsstat.bpf.c` to use fentry.
> So I could not include both SEC() of fentry and kprobes in `vfsstat.bpf.c`.

You can have fentry in your BPF code and if you will do `bpf_program__set_autoload(false)` on it, vmlinux BTF won't be required by libbpf.

Reply:
> > For example, above code will not return 0 and execute libbpf_find_kernel_btf() and fail to find BTF if I change vfsstat.bpf.c to use fentry.
> 
> I understand the mechanism of bpf_program__set_autoload() and it bypasses the check at runtime when false.
> But below code may not be correct. I think that `if (!should_fallback()) {` is correct instead of `if (should_fallback()) {` . Or should_fallback() should return false when BTF is supported.
> 
> https://github.com/iovisor/bcc/blob/7fd8cad30f910af0470e00130817442091376ab3/libbpf-tools/ext4dist.c#L195
> 
> I'll create PR to fix above issue and some minor issues.

I think that code is correct. The difference between what you do and what ext4dist is doing is that ext4dist will always try to use fentry, unless it detects that ext4 is a kernel module and kernel doesn't support kernel module BTF.

We might change it to be still more conservative and fallback to kprobe, if no fentry and vmlinux BTF are supported, but that's a separate issue. As is, the logic is valid: it disables fentry if it is detected that ext4 is a module and module BTF isn't yet supported.